Src/IronPython.Modules/_ctypes/Field.cs

```diff
@@ -112,110 +100,138 @@ public sealed class Field : PythonTypeDataSlot, ICodeFormattable {
             }
 
-#endregion
+            #endregion
 
-#region Private implementation
+            #region Private implementation
 
             /// <summary>
             /// Called for fields which have been limited to a range of bits.  Given the
             /// value for the full type this extracts the individual bits.
             /// </summary>
-            private object ExtractBits(object value) {
-                if (value is int) {
-                    int validBits = ((1 << _bits) - 1);
+            private object ExtractBits(CData instance) {
+                Debug.Assert(_bits > 0);
+                Debug.Assert(_bits < sizeof(ulong) * 8); // Avoid unspecified shifts: ECMA 335: III.3.58-59
+                Debug.Assert(instance != null);
+
+                object value;
+                if (IsBoolType) {
+                    int valueBits = instance.MemHolder.ReadByte(offset);
+                    int ibVal = ExtractBitsFromInt(valueBits);
+                    value = ScriptingRuntimeHelpers.BooleanToObject(ibVal != 0);
 
-                    int iVal = (int)value;
-                    iVal = (iVal >> _bitsOffset) & validBits;
-                    if (IsSignedType) {
-                        // need to sign extend if high bit is set
-                        if ((iVal & (1 << (_bits - 1))) != 0) {
-                            iVal |= (-1) ^ validBits;
-                        }
-                    }
-                    value = ScriptingRuntimeHelpers.Int32ToObject(iVal);
                 } else {
-                    Debug.Assert(value is BigInteger); // we only return ints or big ints from GetValue
-                    ulong validBits = (1UL << _bits) - 1;
+                    value = _fieldType.GetValue(instance.MemHolder, instance, _offset, false);
+                    if (value is int iVal) {
+                        iVal = ExtractBitsFromInt(iVal);
+                        value = ScriptingRuntimeHelpers.Int32ToObject(iVal);
+                    } else if (value is BigInteger biVal) {
+                        ulong validBits = (1UL << _bits) - 1;
+
+                        ulong bits;
+                        if (IsSignedType) {
+                            bits = (ulong)(long)biVal;
+                        } else {
+                            bits = (ulong)biVal;
+                        }
+
+                        bits = (bits >> _bitsOffset) & validBits;
+
+                        if (IsSignedType) {
+                            // need to sign extend if high bit is set
+                            if ((bits & (1UL << (_bits - 1))) != 0) {
+                                bits |= ulong.MaxValue ^ validBits;
+                            }
+                            value = (BigInteger)(long)bits;
+                        } else {
+                            value = (BigInteger)bits;
+                        }
+
+                    } else if (value is bool bVal) {
```

Comment:
Is this reachable or `bool` will be caught by the `IsBoolType` check above?

Reply:
No, not reachable. Again a leftover from a previous version.